### PR TITLE
[Enhancement] decimal multiplication opt

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function
Implement fast decimal multiplication for cases following:
1. decimal32 x decimal32 produces decimal64/decimal128 directly without casting decimal32 to decimal64/decimal128.
2. decimal32 x decimal64 produces decimal64 directly without casting decimal32/decimal64 to decimal decimal128.
3. decimal64 x decimal64 produces decimal128 directly without casting decimal64 to decimal128.

## Test

Several tests have be conducted on TPCDS-10T. table store_sales(28800104574) are used.

**Test queries**
```
 Q1(i64xi64=i64) : select sum((ss_wholesale_cost * ss_list_price)) from store_sales;
 Q2(i128xi128=i128): select sum(((ss_wholesale_cost+1000000000.00000001) * ss_list_price)) from store_sales;
 Q3(i64xi64=i128): select sum((ss_wholesale_cost + (CAST(1000000000.01 AS DECIMAL64(13,2)))) * ss_list_price) from store_sales;
``` 
Q1 is decimal64 multiplying decimal64 produces decimal64, so it is not optimized.
Q2 is decimal128 multiplying decimal64 produces decimal64, it is also not optimized.
Q3 is decimal64 multiplying decimal64 produces decimal128, it is optimized.

decimal multiplication adopts integer multiplication underlyingly.

i64 x i64 => i128 use inline assembly, reduce 3 64-bit mul instructions to 1.
```cpp
template <class T, class U, std::enable_if_t<is_bit64<T>, T> = 0, std::enable_if_t<is_bit64<U>, U> = 0>
static int asm_mul(T x, U y, Int128Wrapper& res) {
    int8_t overflow;
    if constexpr (std::is_unsigned<T>::value && std::is_unsigned<U>::value) {
        __asm__ __volatile__(
                "mov %[x], %%rax\n\t"
                "mul %[y]\n\t"
                "mov %%rdx, %[high]\n\t"
                "mov %%rax, %[low]"
                : [ high ] "=r"(res.u.high), [ low ] "=r"(res.u.low), "=@cco"(overflow)
                : [ x ] "r"(x), [ y ] "r"(y)
                : "cc", "rdx", "rax");
    } else {
        __asm__ __volatile__(
                "mov %[x], %%rax\n\t"
                "imul %[y]\n\t"
                "mov %%rdx, %[high]\n\t"
                "mov %%rax, %[low]"
                : [ high ] "=r"(res.s.high), [ low ] "=r"(res.s.low), "=@cco"(overflow)
                : [ x ] "r"(x), [ y ] "r"(y)
                : "cc", "rdx", "rax");
    }
    return overflow;
}

```

i32 x i32 =>i64 can also optimized in the same way, but decimal32 is not exposed to our users, the inline assembly codes as follows:
```cpp
static int64_t asm_mul32(int32_t x, int32_t y) {
    union {
        int64_t i64;
        struct {
#if __BYTE_ORDER == LITTLE_ENDIAN
            int32_t low;
            int32_t high;
#else
            int32_t high;
            int32_t low;
#endif
        } s;
    } z;
    __asm__ __volatile__(
            "mov %[x], %%eax\n\t"
            "imul %[y]\n\t"
            "mov %%edx, %[high]\n\t"
            "mov %%eax, %[low]"
            : [ high ] "=r"(z.s.high), [ low ] "=r"(z.s.low)
            : [ x ] "r"(x), [ y ] "r"(y)
            : "cc", "rax", "rdx");
    return z.i64;
}
```

**Test result(pipeline_dop=1)**
``` 
+===================+=================+======================+==================+=======================+=========+
| query             | pre-opt latency | pre-opt project cost | post-opt latency | post-opt project cost | speedup |
+===================+=================+======================+==================+=======================+=========+
| Q1(i64xi64=i64)   | 20.190s         | 5.736s               | 18.795s          | 5.743s                | 1X      |
| Q2(i128xi28=i128) | 102.887s        | 80s                  | 95.574s          | 81s                   | 1X      |
| Q3(i64xi64=i128)  | 69.426s         | 50.705s              | 27.861s          | 14.354s              | 3.54X   |
+-------------------+-----------------+----------------------+------------------+-----------------------+---------+
```
- {pre-opt, post-opt} latency means the entire cost of the query.
- {pre-opt, post-opt} project cost means the PushTotalTime of ProjectOperator of the query, this profile metric can be used to show the cost of the multiplication. 
- speedup means pre-opt  ratio of project cost / post-opt project cost.
- **so we can find that Q3 show the 3.54X speedup  in the terms of project cost.** and in pre-opt test, project cost in Q3 is close to Q2.

**Test result(pipeline_dop=0)**
```
+===================+=================+======================+==================+=======================+=========+
| query             | pre-opt latency | pre-opt project cost | post-opt latency | post-opt project cost | speedup |
+===================+=================+======================+==================+=======================+=========+
| Q1(i64xi64=i64)   | 8.247s          | 1.421s               | 7.923s           | 1.418s                | 1X      |
| Q2(i128xi28=i128) | 21.271s         | 16.699s              | 20.853s          | 17.1s                 | 1X      |
| Q3(i64xi64=i128)  | 15.430s         | 10.908s              | 9.337s           | 3.80s                 | 2.87X   |
+-------------------+-----------------+----------------------+------------------+-----------------------+---------+
```
- **In Q3, post-opt project  outperform pre-opt project 2.87X.**

### CODE REVIEW

File: be/src/runtime/int128_arithmetics_x86_64.h [hunk 1]
```diff
@@ -93,15 +93,40 @@ static int asm_mul(T x, U y, Int128Wrapper& res) {
...
                 "mov %%rax, %[low]"
-                : [ low ] "=r"(res.s.low), "=@cco"(overflow)
+                : [ high ] "=r"(res.s.high), [ low ] "=r"(res.s.low), "=@cco"(overflow)
                 : [ x ] "r"(x), [ y ] "r"(y)
                 : "cc", "rdx", "rax");
     }
     return overflow;
 }
 
+static int64_t asm_mul32(int32_t x, int32_t y) {
```
Comment (stdpain) [inline be/src/runtime/int128_arithmetics_x86_64.h:106]: Is it works on ARM ?

File: be/src/exprs/vectorized/arithmetic_expr.cpp [hunk 2]
```diff
@@ -18,11 +21,68 @@ namespace starrocks::vectorized {
                                                       \
     virtual Expr* clone(ObjectPool* pool) const override { return pool->add(new CLASS_NAME(*this)); }
 
+static std::optional<PrimitiveType> eliminate_trivial_cast_for_decimal_mul(const Expr* e) {
```
Comment (stdpain) [inline be/src/exprs/vectorized/arithmetic_expr.cpp:24]: may be we could define a function in FE ? Mul(decimal32, decimal128) ?

Comment (satanson) [reply be/src/exprs/vectorized/arithmetic_expr.cpp:24]: It is not required, we can consider its an abstract leak to optimize general-purpose multiplication.

File: be/src/runtime/int128_arithmetics_x86_64.h (hunk 1 above)
Comment (satanson) [reply be/src/runtime/int128_arithmetics_x86_64.h:106]: asm_mul32 is enclosed by X86_64 macro finally. this case is not optimized on arm.

File: be/src/runtime/int128_arithmetics_x86_64.h [hunk 3]
```diff
@@ -93,15 +93,40 @@ static int asm_mul(T x, U y, Int128Wrapper& res) {
...
+        struct {
+#if __BYTE_ORDER == LITTLE_ENDIAN
+            int32_t low;
+            int32_t high;
+#else
+            int32_t high;
+            int32_t low;
+#endif
+        } s;
+    } z;
```
Comment (LiShuMing) [inline be/src/runtime/int128_arithmetics_x86_64.h:118]: Very clever design. Use a `union` to check overflow.

File: be/test/exprs/vectorized/decimal_binary_function_test.cpp [hunk 4]
```diff
@@ -2837,4 +2955,121 @@ TEST_F(DecimalBinaryFunctionTest, test_decimal128p38s14_div_decimal128p38s14_eq_
...
+            {"5300.1993", "-14319371956134.8862", "-75895525218345754.54281966"},
+            {"52771758641568.0008", "-25416153656949.7962", "-1341255126381560552930609374.00143696"},
+            {"3.2917", "-99999999999999.9213", "-329169999999999.74094321"},
+            {"-78809072751576.0845", "-14758895507471.3475", "1163134869781218860114475849.98886375"},
+            {"-63650347027092.4971", "83952744775493.0151", "-5343621338837057001558324362.66700621"},
+            {"197070.9764", "20234892726480.4061", "3987710066956751765.59551604"},
+            {"0.0114", "-75288876496539.0058", "-858293192060.54466612"},
+            {"-72638466358742.9662", "3.1225", "-226813611205174.91195950"},
+            {"60915079068521.5269", "-99999121308263.0016", "-6091454381275516575941632856.40914304"}};
+    test_decimal_fast_mul_help<TYPE_DECIMAL64, TYPE_DECIMAL64, TYPE_DECIMAL128, MulOp64x64_128>(test_cases, 18, 4, 18,
```
Comment (LiShuMing) [inline be/test/exprs/vectorized/decimal_binary_function_test.cpp:3071]: Are there some overflow cases?

File: be/src/exprs/vectorized/arithmetic_expr.cpp [hunk 5]
```diff
@@ -18,11 +21,68 @@ namespace starrocks::vectorized {
...
+    }
+}
+
 template <PrimitiveType Type, typename OP>
 class VectorizedArithmeticExpr final : public Expr {
 public:
     DEFINE_CLASS_CONSTRUCTOR(VectorizedArithmeticExpr);
+
+    std::optional<ColumnPtr> evaluate_decimal_fast_mul(ExprContext* context, vectorized::Chunk* chunk) {
+        auto lhs_pt_opt = eliminate_trivial_cast_for_decimal_mul(_children[0]);
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/arithmetic_expr.cpp:44]: I think we should do this logic in FE

File: be/src/runtime/int128_arithmetics_x86_64.h [hunk 6]
```diff
@@ -93,15 +93,40 @@ static int asm_mul(T x, U y, Int128Wrapper& res) {
...
+    __asm__ __volatile__(
```
Comment (mofeiatwork) [inline be/src/runtime/int128_arithmetics_x86_64.h:119]: Is there any existed library implement these optimizations ?

File: be/src/exprs/vectorized/cast_expr.cpp [hunk 7]
```diff
@@ -1031,6 +1031,8 @@ class VectorizedCastExpr final : public Expr {
             << ", expr=" << expr_debug_string << ")";
         return out.str();
     }
+
+    bool is_cast_expr() const override { return true; }
```
Comment (mofeiatwork) [inline be/src/exprs/vectorized/cast_expr.cpp:1035]: `Expr::_opcode` could check the expression type. Use this virtual method is meaningless.

File: be/test/exprs/vectorized/decimal_binary_function_test.cpp (hunk 4 above)
Comment (satanson) [reply be/test/exprs/vectorized/decimal_binary_function_test.cpp:3071]: this situation never overflows, try to think, i64 x i64 yields i128, can it overflow?

File: be/src/exprs/vectorized/arithmetic_expr.cpp (hunk 5 above)
Comment (satanson) [reply be/src/exprs/vectorized/arithmetic_expr.cpp:44]: frankly speak, I think it is a not a good idea. arithmetic operation rules are stable to users, how to compute is intrinsic things.  Only modifying code of BE have at lest two benefits.
1. Users are not puzzled by sometimes i64 x i64=>i128,  somtimes i128 x i128 => i128. the DBA would be disturbed by users to tell why.
2. There are so many arithmetic optization cases, for examples, i8 x i8 => float32, i8 x i8 => i16, i16 x i16=>i32, in the future,if we optimize these cases, we just opt code of be.

**Review by stdpain — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**